### ISSUE TITLE
AWS subnets mixed up when creating VPC

### ISSUE THREAD

**Issue by rhartkopf** [ansible/ansible-modules-core#3123]
##### Issue Type:
- Bug Report
##### Plugin Name:

ec2_vpc
##### Ansible Version:

```
ansible 2.0.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### Ansible Configuration:

Defaults
##### Environment:

Ubuntu 14.04.3 LTS
##### Summary:

When creating a VPC, the subnet name tags are created correctly on the first run, then get jumbled up on consecutive runs.
##### Steps To Reproduce:

vars:

```
ec2_region: us-east-1

vpc_name: test_vpc
vpc_cidr_block: 10.0.0.0/16
vpc_subnets:
  - cidr: 10.0.0.0/24
    az: us-east-1a
    resource_tags: { "Name":"VPN" }
  - cidr: 10.0.10.0/24
    az: us-east-1a
    resource_tags: { "Name":"Ops" }
  - cidr: 10.0.20.0/24
    az: us-east-1a
    resource_tags: { "Name":"StagingA" }
  - cidr: 10.0.30.0/24
    az: us-east-1c
    resource_tags: { "Name":"StagingB" }
  - cidr: 10.0.40.0/24
    az: us-east-1d
    resource_tags: { "Name":"ProdA" }
  - cidr: 10.0.50.0/24
    az: us-east-1e
    resource_tags: { "Name":"ProdB" }
vpc_route_tables:
  - subnets:
      - 10.0.0.0/24
      - 10.0.10.0/24
      - 10.0.20.0/24
      - 10.0.30.0/24
      - 10.0.40.0/24
      - 10.0.50.0/24
    routes:
      - dest: 0.0.0.0/0
        gw: igw
```

task:

```
- name: ec2 vpc created
  local_action:
    module: ec2_vpc
    state: present
    cidr_block: "{{ vpc_cidr_block }}"
    resource_tags: '{ "Name":"{{ vpc_name }}" }'
    region: "{{ ec2_region }}"
    subnets: "{{ vpc_subnets }}"
    route_tables: "{{ vpc_route_tables }}"
    internet_gateway: True
  register: vpc
  run_once: true
```
##### Expected Results:

I expected to see subnet tags that match my configuration.
![screen shot 2016-02-25 at 10 31 58](https://cloud.githubusercontent.com/assets/5224897/13328284/0b852c5c-dbab-11e5-8d71-0d64fb31e8b8.png)
##### Actual Results:

The subnets are created with correct AZ and CIDR, but some subnets are tagged incorrectly.

```
TASK [aws : ec2 vpc created] ***************************************************
task path: /masked/roles/aws/tasks/ec2_vpc.yml:4
ESTABLISH LOCAL CONNECTION FOR USER: jenkins
localhost EXEC /bin/sh -c '( umask 22 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1456419924.16-243276569381007 `" && echo "` echo $HOME/.ansible/tmp/ansible-tmp-1456419924.16-243276569381007 `" )'
localhost PUT /tmp/tmpbDtLaL TO /var/lib/jenkins/.ansible/tmp/ansible-tmp-1456419924.16-243276569381007/ec2_vpc
localhost EXEC /bin/sh -c 'LANG=C LC_ALL=C LC_MESSAGES=C /usr/bin/python /var/lib/jenkins/.ansible/tmp/ansible-tmp-1456419924.16-243276569381007/ec2_vpc; rm -rf "/var/lib/jenkins/.ansible/tmp/ansible-tmp-
1456419924.16-243276569381007/" > /dev/null 2>&1'
changed: [ops -> localhost] => {"changed": true, "igw_id": "igw-3d528a59", "invocation": {"module_args": {"aws_access_key": null, "aws_secret_key": null, "cidr_block": "10.0.0.0/16", "
dns_hostnames": true, "dns_support": true, "ec2_url": null, "instance_tenancy": "default", "internet_gateway": true, "profile": null, "region": "us-east-1", "resource_tags": {"Name": "test_vpc"}, "
route_tables": [{"routes": [{"dest": "0.0.0.0/0", "gw": "igw"}], "subnets": ["10.0.0.0/24", "10.0.10.0/24", "10.0.20.0/24", "10.0.30.0/24", "10.0.40.0/24", "10.0.50.0/24"]}], "security_token": null, "stat
e": "present", "subnets": [{"az": "us-east-1a", "cidr": "10.0.0.0/24", "resource_tags": {"Name": "VPN"}}, {"az": "us-east-1a", "cidr": "10.0.10.0/24", "resource_tags": {"Name": "Ops"}}, {"az": "us-east-1a
", "cidr": "10.0.20.0/24", "resource_tags": {"Name": "StagingA"}}, {"az": "us-east-1c", "cidr": "10.0.30.0/24", "resource_tags": {"Name": "StagingB"}}, {"az": "us-east-1d", "cidr": "10.0.40.0/24", "resour
ce_tags": {"Name": "ProdA"}}, {"az": "us-east-1e", "cidr": "10.0.50.0/24", "resource_tags": {"Name": "ProdB"}}], "validate_certs": true, "vpc_id": null, "wait": false, "wait_timeout": 300}, "module_name":
 "ec2_vpc"}, "subnets": [{"az": "us-east-1a", "cidr": "10.0.0.0/24", "id": "subnet-b88c2ce0", "resource_tags": {"Name": "ProdB"}}, {"az": "us-east-1a", "cidr": "10.0.10.0/24", "id": "subnet-bea77ce6", "re
source_tags": {"Name": "ProdB"}}, {"az": "us-east-1a", "cidr": "10.0.20.0/24", "id": "subnet-b98c2ce1", "resource_tags": {"Name": "ProdB"}}, {"az": "us-east-1c", "cidr": "10.0.30.0/24", "id": "subnet-468e
506c", "resource_tags": {"Name": "ProdB"}}, {"az": "us-east-1d", "cidr": "10.0.40.0/24", "id": "subnet-ff9d1489", "resource_tags": {"Name": "ProdB"}}, {"az": "us-east-1e", "cidr": "10.0.50.0/24", "id": "s
ubnet-ae096793", "resource_tags": {"Name": "ProdA"}}], "vpc": {"cidr_block": "10.0.0.0/16", "dhcp_options_id": "dopt-e8f8128d", "id": "vpc-a086d9c4", "region": "us-east-1", "state": "available"}, "vpc_id"
: "vpc-a086d9c4"}
```

![screen shot 2016-02-25 at 10 28 28](https://cloud.githubusercontent.com/assets/5224897/13328187/9b0dd5be-dbaa-11e5-9597-cb6517685ad1.png)

**Comment by bonovoxly** [comment]
I'm seeing similar broken behavior in 2.0.1.0 with regards to subnets.

**Comment by njsaunders** [comment]
Same. The following code:

```
    - name: Create VPC
      ec2_vpc:
        aws_access_key: "{{ assumed_role.sts_creds.access_key }}"
        aws_secret_key: "{{ assumed_role.sts_creds.secret_key }}"
        security_token: "{{ assumed_role.sts_creds.session_token }}"
        state: present
        cidr_block: "{{vpc_octet}}.0.0/16"
        resource_tags: { "environment": "{{envname}}" }
        region: "{{ region }}"
        internet_gateway: yes
        subnets:
          - cidr: "{{vpc_octet}}.0.0/24"
            az: "{{region}}a"
            resource_tags: { "Name": "{{region}}a public" }
          - cidr: "{{vpc_octet}}.10.0/24"
            az: "{{region}}b"
            resource_tags: { "Name": "{{region}}b public" }
          - cidr: "{{vpc_octet}}.20.0/24"
            az: "{{region}}c"
            resource_tags: { "Name": "{{region}}c public" }
          - cidr: "{{vpc_octet}}.1.0/24"
            az: "{{region}}a"
            resource_tags: { "Name": "{{region}}a awsprivate" }
          - cidr: "{{vpc_octet}}.11.0/24"
            az: "{{region}}b"
            resource_tags: { "Name": "{{region}}b awsprivate" }
          - cidr: "{{vpc_octet}}.21.0/24"
            az: "{{region}}c"
            resource_tags: { "Name": "{{region}}c awsprivate" }
          - cidr: "{{vpc_octet}}.2.0/24"
            az: "{{region}}a"
            resource_tags: { "Name": "{{region}}a private" }
          - cidr: "{{vpc_octet}}.12.0/24"
            az: "{{region}}b"
            resource_tags: { "Name": "{{region}}b private" }
          - cidr: "{{vpc_octet}}.22.0/24"
            az: "{{region}}c"
            resource_tags: { "Name": "{{region}}c private" }
        route_tables:
        - subnets:
            - "{{vpc_octet}}.0.0/24"
            - "{{vpc_octet}}.10.0/24"
            - "{{vpc_octet}}.20.0/24"
            - "{{vpc_octet}}.1.0/24"
            - "{{vpc_octet}}.11.0/24"
            - "{{vpc_octet}}.21.0/24"
            - "{{vpc_octet}}.2.0/24"
            - "{{vpc_octet}}.12.0/24"
            - "{{vpc_octet}}.22.0/24"
          routes:
            - dest: 0.0.0.0/0
              gw: igw
      register: vpc_out
```

Produces:

<img width="921" alt="screen shot 2016-03-03 at 10 18 11" src="https://cloud.githubusercontent.com/assets/1478320/13491252/4ab9c832-e129-11e5-9676-d6abb045af62.png">

**Comment by carmi** [comment]
This issue is also occuring for me. It was working properly in 2.0.0.2 and after upgrading to 2.0.1.0 it is breaking (naming all subnets the value of the final Name resource_tags value)

**Comment by jimsmith** [comment]
I to am also seeing this issue, the first time the tags for subnets are correct, however when I re-run the playbook since it's id-impotent for when I was updating some of the subnet Name tags I hit this same issue.

**Comment by jonhadfield** [comment]
Same issue here.
A quick look suggests a regression caused by: https://github.com/ansible/ansible-modules-core/commit/cba7987

**Comment by balor** [comment]
+1

**Comment by diclophis** [comment]
Our OPS department can also confirm this buggy behavior... and would like to note that it is blocking our upgrade to ansible 2.x

**Comment by armandocerna** [comment]
We are also experience this show stopping bug and is also blocking our adoption of ansible 2

**Comment by jessewmc** [comment]
I was also having this issue in 2.0.1. I pulled the latest devel from github to see if I could fix it, but it seems someone already has. Not sure how far back it was fixed, but the latest version on devel head appears to be 2.1, well ahead of current release.

**Comment by dmrickert** [comment]
For reference/explanation, it looks like it was broken by this commit:
https://github.com/ansible/ansible-modules-core/commit/f769305bd4e96893323a965e3fa7ed68ef7dd40a

And then fixed by this commit:
https://github.com/ansible/ansible-modules-core/commit/c30bca68088bf181a895b13660faac0a8f098105

**Comment by diclophis** [comment]
What concerns me the most with these regressions/fixes is that there does not appear to be any level of unit testing occurring at the module level for any of these changes... This concerns me _alot_ ... how can we be assured that changes between ansible versions are compatible/working correctly?

**Comment by dtoubelis** [comment]
Here are some more details on this issue (ansible v2.0.1.0, Ubuntu-14.04):

```
- name: Create VPC
  ec2_vpc:
    state: present
    region: "{{ region }}"
    cidr_block: "10.{{ network_number }}.0.0/16"
    internet_gateway: yes
    resource_tags:
      Name: prod-us-east-1
    subnets:
      - cidr: "10.{{ network_number }}.4.0/22"
        az: us-east-1a
        resource_tags:
          Env: prod
      - cidr: "10.{{ network_number }}.8.0/22"
        az: us-east-1b
        resource_tags:
          Scope: pub
    route_tables:
      - subnets:
        - "10.{{ network_number }}.4.0/22"
        - "10.{{ network_number }}.8.0/22"
        routes:
        - dest: 0.0.0.0/0
          gw: igw
        resource_tags:
          Name: prod-pub
    wait: yes
  register: vpc
```

As you see the subnets have different tag scopes but in amazon console each subnet will have both tags. So, it looks like all tags are applied to all subnets and likely in parallel causing mixups and unpredictable ordering.

**Comment by thaumos** [comment]
Issue #2930 is referencing the same thing.

PR #2931 fixed this issue and #2930

**Comment by dtoubelis** [comment]
Any chance it can be retrofitted into 2.0.x branch and pushed to Ubuntu PPA?

**Comment by jimsmith** [comment]
Do you know when EPEL repo will have this pushed into it ?
as it's presently version 2.0.1.0  in the epel-testing repo

Name        : ansible
Arch        : noarch
Version     : 2.0.1.0
Release     : 2.el6
Size        : 12 M
Repo        : installed
From repo   : epel-testing
Summary     : SSH-based configuration management, deployment, and task execution system
URL         : http://ansible.com
License     : GPLv3+
Description :
            : Ansible is a radically simple model-driven configuration management,
            : multi-node deployment, and remote task execution system. Ansible works
            : over SSH and does not require any software or daemons to be installed
            : on remote nodes. Extension modules can be written in any language and
            : are transferred to managed machines automatically.

**Comment by gaieges** [comment]
+1

**Comment by abadger** [comment]
Thanks everyone.  Closing as fixed in devel (will be 2.1.0) via #2931 

2.1.0 will likely be out before a theoretical 2.0.3 is released.  We'll probably simply encourage people to update to 2.1.0 instead.
